feat/(PROD-CPC-1037): Recently Seen Products

ZacharyLelievre opened the pull request:
[**JIRA:** link to jira ticket
](https://champlainsaintlambert.atlassian.net/browse/CPC-1037)
## Context:
This ticket focused on implementing a 'Recently Clicked' feature that displays each product selected by the user at the bottom of the page. The feature tracks the products a user interacts with during their session and dynamically updates a section to showcase those products.
## Does this PR change the .vscode folder in petclinic-frontend?:
The PR does not affect the .vscode

## Changes

- State management: Added useState in ProductList to track the list of recently clicked products.
- Functionality: Implemented a function that updates the list of clicked products and ensures it displays properly on the page.

## Before and After UI (Required for UI-impacting PRs)

Before:
![image](https://github.com/user-attachments/assets/5ee02cd5-c8f3-4b5d-a2a0-bb71e9da176f)


After:
![image](https://github.com/user-attachments/assets/0aba50bb-c3dd-4bea-b44f-ac3add98edad)

## Dev notes (Optional)
Specific technical changes that should be noted
## Linked pull requests (Optional)
Pull request links

Valthefirst commented:
When you tested your feature before pushing did you check that it intergrated well with the other features. I don't see the trending section.

J91-cloud reviewed (changes requested):
Zachary, It is well done but i recommend that when clicking on new trending products the list is in chronological order. 

For example: 

![Screenshot 2024-09-25 154228](https://github.com/user-attachments/assets/82acb77f-4a56-4ab5-b85c-a59bd6c01141)

Horse Saddle is the last one i clicked but it appears to be first in the list. When i click on a new item it would be added to the right. Next to Bird Cage on the right. 

Maybe you can revert the list.

viktorkuts commented:
Looks good, but is there a reason why you create your own implementation for the Product when you could've re-used the Product component which is utilized in ProductList?

ThomasBedard commented:
<img width="317" alt="Screenshot 2024-09-25 at 3 50 41 PM" src="https://github.com/user-attachments/assets/e91aa8f8-50e8-4d68-9f39-66a589d98516">
<img width="825" alt="Screenshot 2024-09-25 at 3 50 49 PM" src="https://github.com/user-attachments/assets/9216e138-aff5-4283-9114-3da016af5669">
<img width="918" alt="Screenshot 2024-09-25 at 3 50 59 PM" src="https://github.com/user-attachments/assets/61be510c-cfb6-4f0d-bb6c-b5b3694ed7f4">

I'm having multiple issue : 
An added product can't be added to recently seen products
When adding a product to recently view, the clicked card changes styles

J91-cloud reviewed (commented):
Zachary, i recommend to maybe enhance this feature. Right now it seems that we are getting the recently clicked products for everyone. It seems to work, because we are testing the web application locally but when this a publicized application  all users will be interfering with this feature. 

Maybe we you can add a endpoint in the back-end that gets all recently clicked with a userId in the URI. This would make it so everyone has its own recently seen product. rather then everyone. 

Please advise me if i am understanding this correctly.

ThomasBedard commented on petclinic-frontend/src/features/products/ProductsList.tsx:
```diff
@@ -82,6 +85,26 @@ export default function ProductList(): JSX.Element {
+    setRecentlyClickedProducts(listOfProducts => {
+      const updatedProducts = [];
+
+      for (const p of listOfProducts) {
+        if (p.productId !== product.productId) {
+          updatedProducts.push(p);
+        }
+      }
```
```suggestion
      for (const currentProduct of listOfProducts) {
        if (currentProduct.productId !== product.productId) {
          updatedProducts.push(currentProduct);
        }
      }
```

maybe consider making it more readable with changing "p" to currentProduct

ZacharyLelievre replied:
Good point! I've updated p to currentProduct to make the code more readable. Thanks for the suggestion!

Elvis-elvis reviewed (approved):
It looks good but the recent clicked product appears last in the list while I think it should appear first otherwise I like it

Valthefirst commented:
Why when I refresh the list dissapears?

ZacharyLelievre commented:
> Looks good, but is there a reason why you create your own implementation for the Product when you could've re-used the Product component which is utilized in ProductList?

You're absolutely right! I hadn't thought of doing it that way. It's now implemented, and I also realized the product ratings were missing from the clicked items, but that's been fixed as well.

ZacharyLelievre commented:
> Zachary, It is well done but i recommend that when clicking on new trending products the list is in chronological order.
> 
> For example:
> 
> ![Screenshot 2024-09-25 154228](https://private-user-images.githubusercontent.com/118135201/370874234-82acb77f-4a56-4ab5-b85c-a59bd6c01141.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNzM4ODYsIm5iZiI6MTcyNzM3MzU4NiwicGF0aCI6Ii8xMTgxMzUyMDEvMzcwODc0MjM0LTgyYWNiNzdmLTRhNTYtNGFiNS1iODVjLWE1OWJkNmMwMTE0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkyNlQxNzU5NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yYzRhMmU1ZWRhMzAxYjczMjc2YzY0NmUwNjlmNTdhNTc0M2VmOTNmMWQ0YjYwODllNmM5NDljN2ViMThhYzYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.UqoOhpPTL_bR4mZa2l3FtBlqGK7LIbhEVbBxnKr4kQs)
> 
> Horse Saddle is the last one i clicked but it appears to be first in the list. When i click on a new item it would be added to the right. Next to Bird Cage on the right.
> 
> Maybe you can revert the list.

Thank you for the suggestion! I’ve updated the implementation so that new products are now added in chronological order. Initially, I was using .push(), which added the most recent product to the end of the list. I’ve switched to using .unshift() so the newest clicked item now appears at the start of the list, making the most recent clicked product always displayed first.

ZacharyLelievre commented:
> When you tested your feature before pushing did you check that it intergrated well with the other features. I don't see the trending section and the search bar in the screenshots.

I didn’t include the top section in the original screenshot, but here it is with the feature working:
![image](https://github.com/user-attachments/assets/7d1e36cd-3c9c-4f04-bc3c-44321716cd6d)

Regarding the issue you mentioned about refreshing the page, I’ve resolved it by using localstorage to store each clicked product on the user’s computer.

As for the owner issue, I haven’t encountered that problem on my end. It seems to be working correctly for all roles. Let me know if you’re still experiencing it.

ZacharyLelievre commented:
> Do you know why your feature doesn't work when I login as an owner?

As for the owner issue, I haven’t encountered that problem on my end. It seems to be working correctly for all roles. Let me know if you’re still experiencing it.

ZacharyLelievre commented:
> Zachary, i recommend to maybe enhance this feature. Right now it seems that we are getting the recently clicked products for everyone. It seems to work, because we are testing the web application locally but when this a publicized application all users will be interfering with this feature.
> 
> Maybe we you can add a endpoint in the back-end that gets all recently clicked with a userId in the URI. This would make it so everyone has its own recently seen product. rather then everyone.
> 
> Please advise me if i am understanding this correctly.

Actually, since I'm using localStorage, each user will already have their own recently clicked products stored locally, so this issue should be resolved. We can still look into adding an endpoint in the future if needed for next Sprint!

ZacharyLelievre commented:
> <img alt="Screenshot 2024-09-25 at 3 50 41 PM" width="317" src="https://private-user-images.githubusercontent.com/90119345/370876465-e91aa8f8-50e8-4d68-9f39-66a589d98516.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3MzczMzMsIm5iZiI6MTcyNzczNzAzMywicGF0aCI6Ii85MDExOTM0NS8zNzA4NzY0NjUtZTkxYWE4ZjgtNTBlOC00ZDY4LTlmMzktNjZhNTg5ZDk4NTE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDIyNTcxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0MWE1YmM4OThlMjJiMDhkMDYyODMxN2FhZjE4ZTczMDFkMThhNzczMTNiYmZmMzY1YjFlMDdlY2JmMjM5NDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._W2oXWJQWXmaCFWNdpWojXvVGzhHRRFRMuDUBJYWr7E"> <img alt="Screenshot 2024-09-25 at 3 50 49 PM" width="825" src="https://private-user-images.githubusercontent.com/90119345/370876466-9216e138-aff5-4283-9114-3da016af5669.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3MzczMzMsIm5iZiI6MTcyNzczNzAzMywicGF0aCI6Ii85MDExOTM0NS8zNzA4NzY0NjYtOTIxNmUxMzgtYWZmNS00MjgzLTkxMTQtM2RhMDE2YWY1NjY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDIyNTcxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmN2QzYmYxMDAwZjg3MTU2YzIyMjMwODI3YTg1MzYyZjU4YWE5OTRlOWVjNjNkMGM4NDc5M2M4MWZmYjZmYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FzGR99qZVg4OATIb8XdvVwmADhGh3KezrfO5j0UqVtc"> <img alt="Screenshot 2024-09-25 at 3 50 59 PM" width="918" src="https://private-user-images.githubusercontent.com/90119345/370876468-61be510c-cfb6-4f0d-bb6c-b5b3694ed7f4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3MzczMzMsIm5iZiI6MTcyNzczNzAzMywicGF0aCI6Ii85MDExOTM0NS8zNzA4NzY0NjgtNjFiZTUxMGMtY2ZiNi00ZjBkLWJiNmMtYjViMzY5NGVkN2Y0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDIyNTcxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1ZDA1ODNlZGQ5ZDNiMDRjYmM2Yzc4YzhlMzc2NjRmMjhhNmNlYTQ3YmUxZGRmM2ExNzgyYmM4ZTUyZmQ4MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pu6etLxYGKqkYGWXZIKWjE6EFKJTZMlEHoYMxjAnBPc">
> I'm having multiple issues : An added product can't be added to recently seen products When adding a product to recently view, the clicked card changes styles

That’s a good thing you pointed out, but I just wanted to clarify that the issue you're experiencing is actually due to how the requestCount function is working. When a product is added and clicked, the requestCount gets triggered, but it’s initially set to null, which causes a 500 error. This isn't a problem with this PR itself but rather with how the requestCount handles null values which I did not touch.

J91-cloud reviewed (approved):
Great! Thanks for fixing the order of which we see the recently clicked products.

ThomasBedard reviewed (approved):
Fixes were applied, works well 👍
